# PYTHON-10. Введение в Pandas

## 1. Введение

### Импорт библиотеки Pandas

In [181]:
import pandas as pd

### Проверка версии библиотеки

In [182]:
pd.__version__

'1.5.2'

In [183]:
print(pd.__name__)

pandas


## 2. Pandas.Series

### Создание Series

* Способ 1 — из списка с использованием параметров функции pd.Series():

In [184]:
countries = pd.Series(
    data = ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    index = ['UK', 'CA', 'US', 'RU', 'UA', 'BY', 'KZ'],
    name = 'countries'
)
display(countries)

UK       Англия
CA       Канада
US          США
RU       Россия
UA      Украина
BY     Беларусь
KZ    Казахстан
Name: countries, dtype: object

* Способ 2 — из словаря, в котором ключами являются будущие метки, а значениями — будущие значения Series, при этом использование параметра name также возможно:

In [185]:
countries = pd.Series({
    'UK': 'Англия',
    'CA': 'Канада',
    'US': 'США',
    'RU': 'Россия',
    'UA': 'Украина',
    'BY': 'Беларусь',
    'KZ': 'Казахстан'},
    name = 'countries'
)
display(countries)

UK       Англия
CA       Канада
US          США
RU       Россия
UA      Украина
BY     Беларусь
KZ    Казахстан
Name: countries, dtype: object

### Доступ к данным

Доступ к элементам осуществляется с использованием loc или iloc.
Для получения названия страны по коду "US" можно выполнить следующий код:

In [186]:
print(countries.loc['US'])

США


Для того чтобы достать информацию по нескольким индексам, необходимо обернуть интересующие индексы в список:

In [187]:
print(countries.loc[['US', 'RU', 'UK']])

US       США
RU    Россия
UK    Англия
Name: countries, dtype: object


Для получения элемента по индексу "KZ" нужно обратиться через .iloc по номеру 6:

In [188]:
print(countries.iloc[6])

Казахстан


Получим срез из исходной Series с первого по третий элемент:

In [189]:
print(countries.iloc[1:4])

CA    Канада
US       США
RU    Россия
Name: countries, dtype: object


На самом деле loc и iloc можно опустить и обращаться к элементам Series напрямую по индексам, например countries[['UK', 'US', 'UA']] или countries[[0, 2, 4]].

### Задание 2.4

В аптеку поступают партии лекарств. Их названия находятся в списке names, количество единиц товара находится в списке counts.

Например:

names=['chlorhexidine', 'cyntomycin', 'afobazol']
counts=[15, 18, 7]

Напишите функцию create_medications(names, counts), создающую Series medications, индексами которого являются названия лекарств names, а значениями — их количество в партии counts.

Также напишите функцию get_percent(medications, name), которая возвращает долю товара с именем name от общего количества товаров в партии в процентах.

In [190]:
import pandas as pd

def create_medications(names, counts):
    medications = pd.Series(index=names, data=counts, name='Series medications')
    return medications


def get_percent(medications, name):
    return medications[name]/sum(medications)*100 


if __name__ == '__main__':
    names=['chlorhexidine', 'cyntomycin', 'afobazol']
    counts=[15, 18, 7]
    medications = create_medications(names, counts)
    print("medications series: ", medications, sep='\n')
    print("chlorhexidine's part: ", get_percent(medications, "chlorhexidine"))

medications series: 
chlorhexidine    15
cyntomycin       18
afobazol          7
Name: Series medications, dtype: int64
chlorhexidine's part:  37.5


## 3. Pandas.DataFrame

### Создание DATAFRAME

DataFrame создаётся с помощью функции pd.DataFrame(). Так же, как и для Series, для создания объектов DataFrame есть несколько способов:

* Способ 1. Самый простой способ создания DataFrame — из словаря, ключами которого являются имена столбцов будущей таблицы, а значениями — списки, в которых хранится содержимое этих столбцов:

In [191]:
countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'square': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})

display(countries_df)

country  population    square
0     Англия       56.29    133396
1     Канада       38.05   9984670
2        США      322.28   9826630
3     Россия      146.24  17125191
4    Украина       45.50    603628
5   Беларусь        9.50    207600
6  Казахстан       17.04   2724902

Обратите внимание, что, так как мы не задали метки (индексы) DataFrame, они были сгенерированы автоматически. Исправим это, задав индексы вручную:

In [192]:
countries_df.index = ['UK', 'CA', 'US', 'RU', 'UA', 'BY', 'KZ']

display(countries_df)

country  population    square
UK     Англия       56.29    133396
CA     Канада       38.05   9984670
US        США      322.28   9826630
RU     Россия      146.24  17125191
UA    Украина       45.50    603628
BY   Беларусь        9.50    207600
KZ  Казахстан       17.04   2724902

* Способ 2. Также DataFrame можно создать из вложенного списка, внутренние списки которого будут являться строками новой таблицы:

In [193]:
countries_df = pd.DataFrame(
    data = [
        ['Англия', 56.29, 133396],
        ['Канада', 38.05, 9984670],
        ['США', 322.28, 9826630],
        ['Россия', 146.24, 17125191],
        ['Украина', 45.5, 603628],
        ['Беларусь', 9.5, 207600],
        ['Казахстан', 17.04, 2724902]
    ],
    columns= ['country', 'population', 'square'],
    index = ['UK', 'CA', 'US', 'RU', 'UA', 'BY', 'KZ']
)

display(countries_df)

country  population    square
UK     Англия       56.29    133396
CA     Канада       38.05   9984670
US        США      322.28   9826630
RU     Россия      146.24  17125191
UA    Украина       45.50    603628
BY   Беларусь        9.50    207600
KZ  Казахстан       17.04   2724902

### AXIS в DATAFRAME

Рассмотрим разницу в результатах работы методов в зависимости от параметра axis на примере использования метода DataFrame mean() — вычисление среднего по таблице.

Считаем среднее по строкам (axis = 0) в каждом столбце:

In [194]:
countries_df.mean(axis=0)

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_23404\1904880919.py:1: FutureWarning: The default value of numeric_only in DataFrame.mean is deprecated. In a future version, it will default to False. In addition, specifying 'numeric_only=None' is deprecated. Select only valid columns or specify the value of numeric_only to silence this warning.
  countries_df.mean(axis=0)


population    9.070000e+01
square        5.800860e+06
dtype: float64

Считаем среднее по столбцам (axis = 1) в каждой строке:

In [195]:
countries_df.mean(axis=1)

C:\Users\artem.truschelev\AppData\Local\Temp\ipykernel_23404\555943002.py:1: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  countries_df.mean(axis=1)


UK      66726.145
CA    4992354.025
US    4913476.140
RU    8562668.620
UA     301836.750
BY     103804.750
KZ    1362459.520
dtype: float64

### Доступ к данным в DATAFRAME

Доступ к столбцу можно получить разными способами:

* Можно обратиться к DataFrame по имени столбца через точку (однако использование такого способа возможно только тогда, когда имя столбца указано без пробелов):

In [196]:
countries_df.population

UK     56.29
CA     38.05
US    322.28
RU    146.24
UA     45.50
BY      9.50
KZ     17.04
Name: population, dtype: float64

* Другой вариант — обратиться к DataFrame по индексу и указать имя столбца:

In [197]:
countries_df['population']

UK     56.29
CA     38.05
US    322.28
RU    146.24
UA     45.50
BY      9.50
KZ     17.04
Name: population, dtype: float64

Для того чтобы получить доступ к ячейкам таблицы, используются уже знакомые нам loc и iloc.

Получим площадь Великобритании:

In [198]:
countries_df.loc['UK', 'square']

133396

Получим население и площадь, соответствующие России:

In [199]:
countries_df.loc['RU', ['population', 'square']]

population      146.24
square        17125191
Name: RU, dtype: object

Сделаем вырезку из таблицы и получим информацию о населении и площади, соответствующую Украине, Беларуси и Казахстану:

In [200]:
countries_df.loc[['UA', 'BY', 'KZ'],['population', 'square']]

population   square
UA       45.50   603628
BY        9.50   207600
KZ       17.04  2724902

или

In [201]:
countries_df.iloc[4:8, 1:3]

population   square
UA       45.50   603628
BY        9.50   207600
KZ       17.04  2724902

### Задание 3.5

Вы работаете аналитиком в компании ScienceYou. Ваша задача — проанализировать чистую прибыль.

Доходы (incomes), расходы (expenses) и годы (years), соответствующие им, предоставлены вам в виде списков.

Например:

incomes = [478, 512, 196]
expenses = [156, 130, 270]
years = [2018, 2019, 2020]

Создайте функцию create_companyDF(incomes, expenses, years), которая  возвращает DataFrame, составленный из входных данных со столбцами Incomes и Expenses и индексами, соответствующими годам рассматриваемого периода. Пример такого DataFrame представлен ниже.

	    Incomes Expenses
2018	478	    156
2019	512	    130
2020	196	    270

Также напишите функцию get_profit(df, year), которая возвращает разницу между доходом и расходом, записанными в таблице df, за год year. Учтите, что если информация за запрашиваемый год не указана в вашей таблице, вам необходимо вернуть None.

Для проверки вхождения запрашиваемого года в перечень индексов таблицы вам поможет атрибут df.index, который возвращает список индексов таблицы.

In [202]:
import pandas as pd

def create_companyDF(incomes, expenses, years):
    df = pd.DataFrame({
    'incomes': incomes,
    'expenses': expenses,
    },
    index = years
    )
    
    return df

def get_profit(df, year):
    if year in df.index:
        profit = df.loc[year, 'incomes'] - df.loc[year, 'expenses']
        return profit
    else:
        return None



if __name__ == '__main__':
    expenses = [156, 130, 270]
    incomes = [478, 512, 196]
    years = [2018, 2019, 2020]
    df = create_companyDF(incomes, expenses, years)
    year = 2018
    
    print(create_companyDF(incomes, expenses, years))
    print('profit in ', year, 'th: ', get_profit(df, year))


      incomes  expenses
2018      478       156
2019      512       130
2020      196       270
profit in  2018 th:  322


## 4. Работа с различными источниками данных в Pandas

### ЗАПИСЬ В CSV-ФАЙЛ

Экспорт данных в формат csv осуществляется с помощью метода DataFrame to_csv(). 

Основные параметры метода DataFrame to_csv():
* path_or_buf — путь до файла, в который будет записан DataFrame (например, data/my_data.csv);
* sep — разделитель данных в выходном файле (по умолчанию ',');
* decimal — разделитель чисел на целую и дробную части в выходном файле (по умолчанию '.');
* columns — список столбцов, которые нужно записать в файл (по умолчанию записываются все столбцы);
* index — параметр, определяющий, требуется ли создавать дополнительный столбец с индексами строк в файле (по умолчанию True).

Заранее создадим папку data в директории, где лежит наш ноутбук. Теперь давайте сохраним наш DataFrame с информацией о странах в csv-файл countries.csv и положим файл в папку data. При этом укажем, что разделителем в нашем файле будет являться символ ';', а также то, что нам не нужен дополнительный столбец с индексами строк:

In [203]:
countries_df = pd.DataFrame({
    'country': ['Англия', 'Канада', 'США', 'Россия', 'Украина', 'Беларусь', 'Казахстан'],
    'population': [56.29, 38.05, 322.28, 146.24, 45.5, 9.5, 17.04],
    'square': [133396, 9984670, 9826630, 17125191, 603628, 207600, 2724902]
})

countries_df.to_csv('data/countries.csv', index=False, sep=';')

### ЧТЕНИЕ CSV-ФАЙЛА

Для чтения таблицы из csv-файла используется функция модуля Pandas read_csv. Функция возвращает DataFrame и имеет несколько важных параметров.

Основные параметры функции read_csv():
* filepath_or_buffer — путь до файла, который мы читаем;
* sep — разделитель данных (по умолчанию ',');
* decimal — разделитель чисел на целую и дробную часть в выходном файле (по умолчанию '.');
* names — список с названиями столбцов для чтения;
* skiprows — количество строк в файле, которые нужно пропустить (например, файл может содержать служебную информацию, которая нам не нужна).

In [204]:
countries_data = pd.read_csv('data/countries.csv', sep=';')
display(countries_data)

country  population    square
0     Англия       56.29    133396
1     Канада       38.05   9984670
2        США      322.28   9826630
3     Россия      146.24  17125191
4    Украина       45.50    603628
5   Беларусь        9.50    207600
6  Казахстан       17.04   2724902

### ЧТЕНИЕ CSV-ФАЙЛА ПО ССЫЛКЕ

На самом деле файл с данными не обязательно должен храниться у вас на компьютере. Если он находится в открытом доступе по ссылке (например, на Google Диске или GitHub), его можно прочитать и из интернета — для этого достаточно в функции read_csv() вместо пути до файла указать ссылку на файл. Например:

data = pd.read_csv('https://raw.githubusercontent.com/esabunor/MLWorkspace/master/melb_data.csv')
display(data)

### ЗАПИСЬ И ЧТЕНИЕ В ДРУГИХ ФОРМАТАХ

Методы для записи таблиц в файлы отличных от csv форматов:
* to_excel() — запись DataFrame в формат Excel-таблицы (.xlsx);
* to_json() — запись DataFrame в формат JSON (.json);
* to_xml() — запись DataFrame в формат XML-документа (.xml);
* to_sql() — запись DataFrame в базу данных SQL (для реализации этого метода необходимо установить соединение с базой данных).


Методы для чтения таблиц из файлов в отличных от csv форматах:
* read_excel() — чтение из формата Excel-таблицы (.xlsx) в DataFrame;
* read_json() — чтение из формата JSON (.json) в DataFrame;
* read_xml() — чтение из формата XML-документа (.xml) в DataFrame;
* read_sql() — чтение из базы данных SQL в DataFrame (также необходимо установить соединение с базой данных).

## 5. Знакомимся с данными: недвижимость

Данные представляют собой таблицу, в которой содержится 23 столбца:
index — номер строки
Suburb — наименование пригорода
Address — адрес
Rooms — количество комнат в помещении
Type — тип здания (h — дом, коттедж, вилла, терраса; u — блочный, дуплексный дом; t — таунхаус)
Price — цена помещения
Method — метод продажи 
SellerG — риэлторская компания
Date — дата продажи (в формате день/месяц/год)
Distance — расстояния до объекта от центра Мельбурна 
Postcode — почтовый индекс
Bathroom — количество ванных комнат
Car — количество парковочных мест
Landsize — площадь прилегающей территории
BuildingArea — площадь здания
YearBuilt — год постройки
CouncilArea — региональное управление
Lattitude — географическая широта
Longitude — географическая долгота
Regionname — наименование района Мельбурна
Propertycount — количество объектов недвижимости в районе
Coordinates — широта и долгота, объединённые в кортеж

In [205]:
import pandas as pd

melb_data = pd.read_csv('data/melb_data.csv', sep=',')

Какова цена объекта недвижимости под индексом 15?

In [206]:
print(melb_data.iloc[15]['Price'])

1310000.0


Когда был продан объект под индексом 90?

In [207]:
print(melb_data.iloc[90]['Date'])

10/09/2016


Во сколько раз площадь прилегающей территории, на которой находится здание с индексом 3521, больше площади участка, на котором находится здание с индексом 1690? Ответ округлите до целого числа.

In [208]:
print(round(melb_data.iloc[3521]['Landsize'] / melb_data.iloc[1690]['Landsize'], 0))

3.0


## 6. Исследование структуры DataFrame

### ВЫВОД ПЕРВЫХ И ПОСЛЕДНИХ СТРОК

Для этого у DataFrame есть методы head() и tail(), которые возвращают n первых и n последних строк таблицы соответственно (по умолчанию n = 5).

Выведем первые пять строк нашей таблицы:

In [209]:
display(melb_data.head())

index      Suburb           Address  Rooms Type      Price Method SellerG  \
0      0  Abbotsford      85 Turner St      2    h  1480000.0      S  Biggin   
1      1  Abbotsford   25 Bloomburg St      2    h  1035000.0      S  Biggin   
2      2  Abbotsford      5 Charles St      3    h  1465000.0     SP  Biggin   
3      3  Abbotsford  40 Federation La      3    h   850000.0     PI  Biggin   
4      4  Abbotsford       55a Park St      4    h  1600000.0     VB  Nelson   

        Date  Distance  ...  Car  Landsize  BuildingArea  YearBuilt  \
0  3/12/2016       2.5  ...  1.0     202.0         126.0     1970.0   
1  4/02/2016       2.5  ...  0.0     156.0          79.0     1900.0   
2  4/03/2017       2.5  ...  0.0     134.0         150.0     1900.0   
3  4/03/2017       2.5  ...  1.0      94.0         126.0     1970.0   
4  4/06/2016       2.5  ...  2.0     120.0         142.0     2014.0   

   CouncilArea  Lattitude  Longtitude             Regionname  Propertycount  \
0        Yarra   -37.7996    144.9984  Northern Metropolitan         4019.0   
1        Yarra   -37.8079    144.9934  Northern Metropolitan         4019.0   
2        Yarra   -37.8093    144.9944  Northern Metropolitan         4019.0   
3        Yarra   -37.7969    144.9969  Northern Metropolitan         4019.0   
4        Yarra   -37.8072    144.9941  Northern Metropolitan         4019.0   

          Coordinates  
0  -37.7996, 144.9984  
1  -37.8079, 144.9934  
2  -37.8093, 144.9944  
3  -37.7969, 144.9969  
4  -37.8072, 144.9941  

[5 rows x 23 columns]

Следующий код выведет семь последних строк нашей таблицы:

In [210]:
melb_data.tail(7)

index         Suburb        Address  Rooms Type      Price Method  \
13573  13573       Werribee    5 Nuragi Ct      4    h   635000.0      S   
13574  13574    Westmeadows     9 Black St      3    h   582000.0      S   
13575  13575  Wheelers Hill   12 Strada Cr      4    h  1245000.0      S   
13576  13576   Williamstown  77 Merrett Dr      3    h  1031000.0     SP   
13577  13577   Williamstown    83 Power St      3    h  1170000.0      S   
13578  13578   Williamstown   96 Verdon St      4    h  2500000.0     PI   
13579  13579     Yarraville     6 Agnes St      4    h  1285000.0     SP   

             SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
13573  hockingstuart  26/08/2017      14.7  ...  1.0     662.0         172.0   
13574            Red  26/08/2017      16.5  ...  2.0     256.0         126.0   
13575          Barry  26/08/2017      16.7  ...  2.0     652.0         126.0   
13576       Williams  26/08/2017       6.8  ...  2.0     333.0         133.0   
13577          Raine  26/08/2017       6.8  ...  4.0     436.0         126.0   
13578        Sweeney  26/08/2017       6.8  ...  5.0     866.0         157.0   
13579        Village  26/08/2017       6.3  ...  1.0     362.0         112.0   

       YearBuilt  CouncilArea  Lattitude  Longtitude  \
13573     1980.0          NaN  -37.89327   144.64789   
13574     1970.0          NaN  -37.67917   144.89390   
13575     1981.0          NaN  -37.90562   145.16761   
13576     1995.0          NaN  -37.85927   144.87904   
13577     1997.0          NaN  -37.85274   144.88738   
13578     1920.0          NaN  -37.85908   144.89299   
13579     1920.0          NaN  -37.81188   144.88449   

                       Regionname  Propertycount           Coordinates  
13573        Western Metropolitan        16166.0  -37.89327, 144.64789  
13574       Northern Metropolitan         2474.0   -37.67917, 144.8939  
13575  South-Eastern Metropolitan         7392.0  -37.90562, 145.16761  
13576        Western Metropolitan         6380.0  -37.85927, 144.87904  
13577        Western Metropolitan         6380.0  -37.85274, 144.88738  
13578        Western Metropolitan         6380.0  -37.85908, 144.89299  
13579        Western Metropolitan         6543.0  -37.81188, 144.88449  

[7 rows x 23 columns]

### РАЗМЕРНОСТЬ ТАБЛИЦЫ

Далее хотелось бы узнать размер таблицы — количество строк и количество столбцов. Это можно сделать с помощью атрибута shape, который возвращает кортеж с количеством строк и столбцов:

In [211]:
melb_data.shape

(13580, 23)

### ПОЛУЧЕНИЕ ИНФОРМАЦИИ О СТОЛБЦАХ

Для того чтобы получить более детальную информацию о столбцах таблицы, можно использовать метод DataFrame info():

In [212]:
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  float64
 12  Bathroom       13580 non-null  float64
 13  Car            13580 non-null  float64
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  float64
 17  CouncilArea    12211 non-null  object 
 18  Lattit

Данный метод выводит:

информацию об индексах;
информацию об общем количестве столбцов;
таблицу, в которой содержится информация об именах столбцов (Column), количестве непустых значений (Non-Null Count) в каждом столбце и типе данных столбца (Dtype), количестве столбцов, в которых используется определённый тип данных;
количество оперативной памяти в мегабайтах, которое тратится на хранение данных.

### ИЗМЕНЕНИЕ ТИПА ДАННЫХ В СТОЛБЦЕ

Если присмотреться внимательнее к выводу метода info(), а конкретнее — к типам данных столбцов, становится понятно, что некоторые признаки кодируются не совсем корректными типами данных.

Чтобы исправить это, можно воспользоваться методом astype(), который позволяет преобразовать тип данных столбца:

In [213]:
melb_data['Car'] = melb_data['Car'].astype('int64')
melb_data['Bedroom'] = melb_data['Bedroom'].astype('int64')
melb_data['Bathroom'] = melb_data['Bathroom'].astype('int64')
melb_data['Propertycount'] = melb_data['Propertycount'].astype('int64')
melb_data['YearBuilt'] = melb_data['YearBuilt'].astype('int64')
melb_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   index          13580 non-null  int64  
 1   Suburb         13580 non-null  object 
 2   Address        13580 non-null  object 
 3   Rooms          13580 non-null  int64  
 4   Type           13580 non-null  object 
 5   Price          13580 non-null  float64
 6   Method         13580 non-null  object 
 7   SellerG        13580 non-null  object 
 8   Date           13580 non-null  object 
 9   Distance       13580 non-null  float64
 10  Postcode       13580 non-null  int64  
 11  Bedroom        13580 non-null  int64  
 12  Bathroom       13580 non-null  int64  
 13  Car            13580 non-null  int64  
 14  Landsize       13580 non-null  float64
 15  BuildingArea   13580 non-null  float64
 16  YearBuilt      13580 non-null  int64  
 17  CouncilArea    12211 non-null  object 
 18  Lattit

### ПОЛУЧЕНИЕ ОПИСАТЕЛЬНОЙ СТАТИСТИКИ

Часто при работе с таблицей нужно быстро посмотреть на основные статистические свойства её столбцов. Для этого можно воспользоваться методом DataFrame describe().

По умолчанию метод работает с числовыми (int64 и float64) столбцами и показывает число непустых значений (count), среднее (mean), стандартное отклонение (std), минимальное значение (min),  квантили уровней 0.25, 0.5 (медиана) и 0.75 (25%, 50%, 75%) и максимальное значение (max) для каждого столбца исходной таблицы.

Чтобы не увязнуть в обилии информации, выведем на экран значение статистических параметров только для столбцов Distance (расстояние от объекта недвижимости до центра Мельбурна), BuildingArea (площадь здания) и Price (цена объекта):

In [214]:
melb_data.describe().loc[:, ['Distance', 'BuildingArea' , 'Price']]

Distance  BuildingArea         Price
count  13580.000000  13580.000000  1.358000e+04
mean      10.137776    139.633972  1.075684e+06
std        5.868725    392.217403  6.393107e+05
min        0.000000      0.000000  8.500000e+04
25%        6.100000    122.000000  6.500000e+05
50%        9.200000    126.000000  9.030000e+05
75%       13.000000    129.940000  1.330000e+06
max       48.100000  44515.000000  9.000000e+06

На самом деле метод describe() можно применять не только к числовым признакам. С помощью параметра include можно указать тип данных, для которого нужно вывести описательную информацию.

Например, для типа данных object метод describe() возвращает DataFrame, в котором указаны:

* количество непустых строк (count);
* количество уникальных значений (unique);
* самое частое значение — мода —  (top);
* частота — объём использования — этого значения (freq) для каждого столбца типа object исходной таблицы.

In [215]:
melb_data.describe(include=['object'])

Suburb           Address   Type Method SellerG        Date  \
count       13580             13580  13580  13580   13580       13580   
unique        314             13378      3      5     268          58   
top     Reservoir  36 Aberfeldie St      h      S  Nelson  27/05/2017   
freq          359                 3   9449   9022    1565         473   

       CouncilArea             Regionname         Coordinates  
count        12211                  13580               13580  
unique          33                      8               13097  
top       Moreland  Southern Metropolitan  -37.8361, 144.9966  
freq          1163                   4695                  12

### ПОЛУЧЕНИЕ ЧАСТОТЫ УНИКАЛЬНЫХ ЗНАЧЕНИЙ В СТОЛБЦЕ

Для того чтобы определить, сколько раз в столбце повторяется каждый из вариантов значений (т.е. найти частоту для каждого уникального знания), используется метод value_counts().
Данный метод возвращает объект Series, в котором в качестве индексов выступают уникальные категории столбца, а значениями — соответствующая им частота

In [216]:
melb_data['Regionname'].value_counts()

Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: Regionname, dtype: int64

Чтобы сделать вывод более интерпретируемым и понятным, можно воспользоваться параметром normalize. При установке значения этого параметра на True результат будет представляться в виде доли (относительной частоты):

In [217]:
melb_data['Regionname'].value_counts(normalize=True)

Southern Metropolitan         0.345729
Northern Metropolitan         0.286451
Western Metropolitan          0.217084
Eastern Metropolitan          0.108321
South-Eastern Metropolitan    0.033137
Eastern Victoria              0.003903
Northern Victoria             0.003019
Western Victoria              0.002356
Name: Regionname, dtype: float64

### Задание 6.1
Дана таблица data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B']). Что выведет код print(data.shape)?

In [218]:
import pandas as pd

data = pd.DataFrame([[0,1], [1, 0], [1, 1]], columns=['А', 'B'])

print(data.shape)

(3, 2)


### Задание 6.3
Какой тип данных имеет столбец Coordinates?

In [219]:
melb_data['Coordinates'].info()

<class 'pandas.core.series.Series'>
RangeIndex: 13580 entries, 0 to 13579
Series name: Coordinates
Non-Null Count  Dtype 
--------------  ----- 
13580 non-null  object
dtypes: object(1)
memory usage: 106.2+ KB


### Задание 6.5
Сколько пропущенных значений в столбце CouncilArea?

In [220]:
melb_data['CouncilArea'].info()
print('Null Count', 13580 - 12211)

<class 'pandas.core.series.Series'>
RangeIndex: 13580 entries, 0 to 13579
Series name: CouncilArea
Non-Null Count  Dtype 
--------------  ----- 
12211 non-null  object
dtypes: object(1)
memory usage: 106.2+ KB
Null Count 1369


### Задание 6.6
Сколько столбцов после преобразования типов имеет целочисленный тип данных

In [221]:
melb_data.describe(include=['int64'])

index         Rooms      Postcode       Bedroom      Bathroom  \
count  13580.000000  13580.000000  13580.000000  13580.000000  13580.000000   
mean    6789.500000      2.937997   3105.301915      2.914728      1.534242   
std     3920.352663      0.955748     90.676964      0.965921      0.691712   
min        0.000000      1.000000   3000.000000      0.000000      0.000000   
25%     3394.750000      2.000000   3044.000000      2.000000      1.000000   
50%     6789.500000      3.000000   3084.000000      3.000000      1.000000   
75%    10184.250000      3.000000   3148.000000      3.000000      2.000000   
max    13579.000000     10.000000   3977.000000     20.000000      8.000000   

                Car     YearBuilt  Propertycount  
count  13580.000000  13580.000000   13580.000000  
mean       1.611856   1966.788218    7454.417378  
std        0.960793     29.088642    4378.581772  
min        0.000000   1196.000000     249.000000  
25%        1.000000   1960.000000    4380.000000  
50%        2.000000   1970.000000    6555.000000  
75%        2.000000   1975.000000   10331.000000  
max       10.000000   2018.000000   21650.000000

### Задание 6.7
Выберите верные утверждения в отношении таблицы melb_data

Среднее и медианное значения цены дома в таблице отличаются более чем в 10 раз (True or False)

In [222]:
melb_data.describe().loc[:, ['Price']]

Price
count  1.358000e+04
mean   1.075684e+06
std    6.393107e+05
min    8.500000e+04
25%    6.500000e+05
50%    9.030000e+05
75%    1.330000e+06
max    9.000000e+06

Половина проданных зданий находится на расстоянии от 6.1 до 13 километров от центра города (True or False)

In [223]:
melb_data.describe().loc[:, ['Distance']]

Distance
count  13580.000000
mean      10.137776
std        5.868725
min        0.000000
25%        6.100000
50%        9.200000
75%       13.000000
max       48.100000

Средняя площадь здания составляет более 180 квадратных метров (True or False)

In [224]:
melb_data.describe().loc[:, ['BuildingArea']]

BuildingArea
count  13580.000000
mean     139.633972
std      392.217403
min        0.000000
25%      122.000000
50%      126.000000
75%      129.940000
max    44515.000000

В данных есть как минимум одно жилое помещение, которое находится ровно в той точке, которая считается центром Мельбурна  (True or False)

In [ ]:
melb_data.describe().loc[:, ['Distance']]

### Задание 6.8
Выберите верные утверждения в отношении таблицы melb_data

Наименьшее число объектов расположено в районе Southern Metropolitan  (True or False)

In [ ]:
melb_data['Regionname'].value_counts()

Southern Metropolitan         4695
Northern Metropolitan         3890
Western Metropolitan          2948
Eastern Metropolitan          1471
South-Eastern Metropolitan     450
Eastern Victoria                53
Northern Victoria               41
Western Victoria                32
Name: Regionname, dtype: int64

Мельбурн и его пригороды разделены как минимум на восемь районов (True or False)

In [ ]:
melb_data.describe(include=['object']).loc[:, ['Regionname']]

Regionname
count                   13580
unique                      8
top     Southern Metropolitan
freq                     4695

У некоторых различных адресов в таблице совпадают географические координаты (True or False)

In [ ]:
melb_data.describe(include=['object']).loc[:, ['Coordinates']]

Coordinates
count                13580
unique               13097
top     -37.8361, 144.9966
freq                    12

In [ ]:
melb_data.describe(include=['object']).loc[:, ['Address']]

Address
count              13580
unique             13378
top     36 Aberfeldie St
freq                   3

В пригороде Reservoir расположено наибольшее число проданных объектов - 359 (True or False)

In [ ]:
melb_data.describe(include=['object']).loc[:, ['Suburb']]

Suburb
count       13580
unique        314
top     Reservoir
freq          359

### Задание 6.9
Сколько процентов от общего количества домов составляют таунхаусы (тип объекта — t)?

In [ ]:
melb_data['Type'].value_counts(normalize=True)

h    0.695803
u    0.222165
t    0.082032
Name: Type, dtype: float64

## 7. Статистические методы

Агрегирующим в Pandas называется метод, который для каждого столбца возвращает только одно значение — показатель (например, вычисление медианы, максимума, среднего и так далее):

.count() - Количество непустых значений
.mean() - Среднее значение
.min() - Минимальное значение
.max() - Максимальное значение
.var() - Дисперсия
.std() - Стандартное отклонение
.sum() - Сумма
.quantile(x) - Квантиль уровня x
.nunique() - Число уникальных значений

В каждый метод можно передать некоторые параметры, среди которых:

axis  — определяет, подсчитывать параметр по строкам или по столбцам;
numeric_only — определяет, вычислять параметры только по числовым столбцам/строкам или нет (True/False).

Вычислим среднюю цену на объекты недвижимости:

In [ ]:
print(melb_data['Price'].mean())

1075684.079455081


Найдём максимальное количество парковочных мест:

In [ ]:
print(melb_data['Car'].max())

10


А теперь представим, что риэлторская ставка для всех компаний за продажу недвижимости составляет 12%. Найдём общую прибыльность риэлторского бизнеса в Мельбурне. Результат округлим до сотых:

In [ ]:
rate = 0.12
income = melb_data['Price'].sum() * rate
print('Total income of real estate agencies:', round(income, 2))

Total income of real estate agencies: 1752934775.88


Найдём, насколько медианная площадь территории отличается от её среднего значения. Вычислим модуль разницы между медианой и средним и разделим результат на среднее, чтобы получить отклонение в долях:

In [ ]:
landsize_median = melb_data['Landsize'].median() 
landsize_mean =  melb_data['Landsize'].mean()
print(abs(landsize_median - landsize_mean)/landsize_mean)

0.21205713983546193


### МОДАЛЬНОЕ ЗНАЧЕНИЕ

Отдельный интерес представляет статический показатель моды — самого распространённого значения в столбце. Он вычисляется с помощью метода mode().

Модальных значений может быть несколько, то есть несколько значений могут встречаться одинаковое количество раз. Поэтому метод mode(), в отличие от агрегирующих методов, возвращает не одно число, а серию.

In [ ]:
print(melb_data['Rooms'].mode())

0    3
Name: Rooms, dtype: int64


print(melb_data['Rooms'].mode())

In [ ]:
landsize_median = melb_data['BuildingArea'].median() 
landsize_mean =  melb_data['BuildingArea'].mean()
print((abs(landsize_median - landsize_mean)/landsize_mean)*100)

9.764079662364534


In [ ]:
print(melb_data['Bedroom'].mode())

0    3
Name: Bedroom, dtype: int64


## 8. Фильтрация данных в DataFrame

Маской называется Series, которая состоит из булевых значений, при этом значения True соответствуют тем индексам, для которых заданное условие выполняется, в противном случае ставится значение False (например, цена > 2 млн).

Создадим маску и положим её в переменную с именем mask.

In [ ]:
mask = melb_data['Price'] > 2000000
display(mask)

0        False
1        False
2        False
3        False
4        False
         ...  
13575    False
13576    False
13577    False
13578     True
13579    False
Name: Price, Length: 13580, dtype: bool

Для фильтрации нужно просто подставить переменную mask в индексацию DataFrame. Маска показывает, какие строки нужно оставлять в результирующем наборе, а какие — убирать (выведем первые пять строк отфильтрованной таблицы):

In [ ]:
display(melb_data[mask].head())

index       Suburb               Address  Rooms Type      Price Method  \
80     80  Albert Park  112 Beaconsfield Pde      3    h  2850000.0     PI   
85     85  Albert Park     104 Richardson St      4    h  2300000.0      S   
88     88  Albert Park        29 Faussett St      2    h  2120000.0      S   
92     92  Albert Park           2 Dundas Pl      3    h  2615000.0      S   
93     93  Albert Park          23 Finlay St      5    h  2100000.0      S   

     SellerG        Date  Distance  ...  Car  Landsize  BuildingArea  \
80    Buxton   4/03/2017       3.3  ...    0     211.0         198.0   
85  Marshall   7/05/2016       3.3  ...    1     153.0         180.0   
88      Greg  10/09/2016       3.3  ...    1     199.0         107.0   
92    Cayzer  10/12/2016       3.3  ...    1     177.0         181.0   
93      Greg  10/12/2016       3.3  ...    1     237.0         126.0   

    YearBuilt   CouncilArea  Lattitude  Longtitude             Regionname  \
80       1890  Port Phillip   -37.8481    144.9499  Southern Metropolitan   
85       1880  Port Phillip   -37.8447    144.9523  Southern Metropolitan   
88       1900  Port Phillip   -37.8422    144.9554  Southern Metropolitan   
92       1880  Port Phillip   -37.8415    144.9585  Southern Metropolitan   
93       1970  Port Phillip   -37.8436    144.9557  Southern Metropolitan   

    Propertycount         Coordinates  
80           3280  -37.8481, 144.9499  
85           3280  -37.8447, 144.9523  
88           3280  -37.8422, 144.9554  
92           3280  -37.8415, 144.9585  
93           3280  -37.8436, 144.9557  

[5 rows x 23 columns]

Примечание. В результате выполнения фильтрации возвращается новый DataFrame, полученный из исходного, при этом исходная таблица melb_data остаётся без изменений.

Также вовсе не обязательно заносить маску в отдельную переменную — можно сразу вставлять условие в операцию индексации DataFrame, например:

In [ ]:
melb_data[melb_data['Price'] > 2000000]

index          Suburb               Address  Rooms Type      Price  \
80        80     Albert Park  112 Beaconsfield Pde      3    h  2850000.0   
85        85     Albert Park     104 Richardson St      4    h  2300000.0   
88        88     Albert Park        29 Faussett St      2    h  2120000.0   
92        92     Albert Park           2 Dundas Pl      3    h  2615000.0   
93        93     Albert Park          23 Finlay St      5    h  2100000.0   
...      ...             ...                   ...    ...  ...        ...   
13521  13521  Port Melbourne          44 Garton St      4    t  2455000.0   
13523  13523         Prahran        69 Greville St      4    h  2668000.0   
13553  13553    Surrey Hills          20 Albert Cr      4    h  2720000.0   
13555  13555    Surrey Hills              3 Oak St      4    h  3100000.0   
13578  13578    Williamstown          96 Verdon St      4    h  2500000.0   

      Method   SellerG        Date  Distance  ...  Car  Landsize  \
80        PI    Buxton   4/03/2017       3.3  ...    0     211.0   
85         S  Marshall   7/05/2016       3.3  ...    1     153.0   
88         S      Greg  10/09/2016       3.3  ...    1     199.0   
92         S    Cayzer  10/12/2016       3.3  ...    1     177.0   
93         S      Greg  10/12/2016       3.3  ...    1     237.0   
...      ...       ...         ...       ...  ...  ...       ...   
13521     SP  Marshall  26/08/2017       3.5  ...    2     123.0   
13523      S    Biggin  26/08/2017       4.6  ...    2     383.0   
13553      S       Kay  26/08/2017      10.2  ...    2    1005.0   
13555     VB  Marshall  26/08/2017      10.2  ...    3     832.0   
13578     PI   Sweeney  26/08/2017       6.8  ...    5     866.0   

       BuildingArea  YearBuilt   CouncilArea  Lattitude  Longtitude  \
80            198.0       1890  Port Phillip  -37.84810   144.94990   
85            180.0       1880  Port Phillip  -37.84470   144.95230   
88            107.0       1900  Port Phillip  -37.84220   144.95540   
92            181.0       1880  Port Phillip  -37.84150   144.95850   
93            126.0       1970  Port Phillip  -37.84360   144.95570   
...             ...        ...           ...        ...         ...   
13521           0.0       2010           NaN  -37.83349   144.94840   
13523         126.0       1970           NaN  -37.84879   144.98882   
13553         126.0       1920           NaN  -37.82421   145.10352   
13555         126.0       1970           NaN  -37.83564   145.10919   
13578         157.0       1920           NaN  -37.85908   144.89299   

                  Regionname  Propertycount           Coordinates  
80     Southern Metropolitan           3280    -37.8481, 144.9499  
85     Southern Metropolitan           3280    -37.8447, 144.9523  
88     Southern Metropolitan           3280    -37.8422, 144.9554  
92     Southern Metropolitan           3280    -37.8415, 144.9585  
93     Southern Metropolitan           3280    -37.8436, 144.9557  
...                      ...            ...                   ...  
13521  Southern Metropolitan           8648   -37.83349, 144.9484  
13523  Southern Metropolitan           7717  -37.84879, 144.98882  
13553  Southern Metropolitan           5457  -37.82421, 145.10352  
13555  Southern Metropolitan           5457  -37.83564, 145.10919  
13578   Western Metropolitan           6380  -37.85908, 144.89299  

[1007 rows x 23 columns]

Найдём количество зданий с тремя комнатами. Для этого отфильтруем таблицу по условию: обратимся к результирующей таблице по столбцу Rooms и найдём число строк в ней с помощью атрибута shape:

In [ ]:
melb_data[melb_data['Rooms'] == 3].shape[0]

5881

Усложним прошлый пример и найдём число трёхкомнатных домов с ценой менее 300 тысяч:

In [ ]:
melb_data[(melb_data['Rooms'] == 3) & (melb_data['Price'] < 300000)].shape[0]

3

Немного «ослабим» условие: теперь нас будут интересовать дома с ценой менее 300 тысяч, у которых либо число комнат равно 3 либо площадь домов более 100 квадратных метров:

In [ ]:
melb_data[((melb_data['Rooms'] == 3) | (melb_data['BuildingArea'] > 100)) & (melb_data['Price'] < 300000)].shape[1]

23

Фильтрацию часто сочетают со статистическими методами. Давайте найдём максимальное количество комнат в таунхаусах. Так как в результате фильтрации получается DataFrame, то обратимся к нему по столбцу Rooms и найдём максимальное значение:

In [ ]:
melb_data[melb_data['Type'] == 't']['Rooms'].max()

5

А теперь более сложный трюк: найдём медианную площадь здания у объектов, чья цена выше средней. Для того чтобы оградить наш код от нагромождений, предварительно создадим переменную со средней ценой:

In [ ]:
mean_price = melb_data['Price'].mean()
melb_data[melb_data['Price'] > mean_price]['BuildingArea'].median()

126.0

### Задание 8.1
У скольких объектов недвижимости из таблицы melb_data отсутствуют ванные комнаты?

In [ ]:
melb_data[melb_data['Bathroom'] == 0].shape[0]

34

### Задание 8.2
Сколько в таблице melb_data объектов недвижимости, которые были проданы риелторской компанией Nelson и стоимость которых составила больше 3 миллионов?

In [ ]:
melb_data[(melb_data['SellerG'] == 'Nelson') & (melb_data['Price'] > 3000000)].shape[0]

5

### Задание 8.3
Какова минимальная стоимость участка без здания (площадь здания равна 0) в таблице melb_data?

In [ ]:
melb_data[melb_data['BuildingArea'] == 0]['Price'].min()

412500.0

### Задание 8.4
Какова средняя цена объектов недвижимости в таблице melb_data с ценой менее одного миллиона, в которых либо количество комнат больше пяти, либо здание моложе 2015 года?

In [ ]:
melb_data[(melb_data['Price'] < 1000000) & ((melb_data['Rooms'] > 5) | (melb_data['YearBuilt'] > 2015))]['Price'].mean()

769238.6363636364

### Задание 8.5
В каком районе Мельбурна чаще всего продаются виллы и коттеджи (тип здания — h) с ценой меньше трёх миллионов?

In [ ]:
melb_data[(melb_data['Price'] < 3e6) & (melb_data['Type'] == 'h')]['Regionname'].mode()

0    Northern Metropolitan
Name: Regionname, dtype: object

## 9. Закрепление знаний

In [ ]:
import pandas as pd

students_performance = pd.read_csv('data/students_performance.csv', sep=',')

students_performance.head()

gender race/ethnicity parental level of education         lunch  \
0  female        group B           bachelor's degree      standard   
1  female        group C                some college      standard   
2  female        group B             master's degree      standard   
3    male        group A          associate's degree  free/reduced   
4    male        group C                some college      standard   

  test preparation course  math score  reading score  writing score  
0                    none          72             72             74  
1               completed          69             90             88  
2                    none          90             95             93  
3                    none          47             57             44  
4                    none          76             78             75

### Задание 9.1
Данные о скольких студентах содержатся в таблице?

In [ ]:
students_performance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


### Задание 9.2
Каков балл по письму у студента под индексом 155?

In [ ]:
students_performance.iloc[155]['writing score']

88

### Задание 9.3 - 9.5
Сколько суммарно пропущенных значений в таблице?
Сколько столбцов в таблице имеет тип данных object?
Какой объём памяти (в килобайтах) занимает таблица?

In [ ]:
students_performance.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 8 columns):
 #   Column                       Non-Null Count  Dtype 
---  ------                       --------------  ----- 
 0   gender                       1000 non-null   object
 1   race/ethnicity               1000 non-null   object
 2   parental level of education  1000 non-null   object
 3   lunch                        1000 non-null   object
 4   test preparation course      1000 non-null   object
 5   math score                   1000 non-null   int64 
 6   reading score                1000 non-null   int64 
 7   writing score                1000 non-null   int64 
dtypes: int64(3), object(5)
memory usage: 62.6+ KB


### Задание 9.6
Каков у студентов средний балл по математике?

In [ ]:
round(students_performance['math score'].mean(), 0)

66.0

### Задание 9.7
Какая расовая группа является самой крупной в учебном заведении?

In [ ]:
students_performance['race/ethnicity'].mode()

0    group C
Name: race/ethnicity, dtype: object

### Задание 9.8
Каков средний балл по чтению у студентов, которые посещали курсы подготовки к экзаменам?

In [ ]:
round(students_performance[students_performance['test preparation course'] == 'completed']['reading score'].mean(), 0)

74.0

### Задание 9.9
Сколько студентов получили 0 баллов по математике?

In [ ]:
students_performance[students_performance['math score'] == 0].shape[0]

1

### Задание 9.10
Проверьте гипотезу: у студентов с оплачиваемым питанием средний балл по математике выше, чем у студентов с льготным питанием.
В качестве ответа напишите наибольший средний балл по математике среди этих групп студентов.
Округлите ответ до целого числа.

In [ ]:
round(students_performance[students_performance['lunch'] == 'standard']['math score'].mean(), 0)

70.0

In [ ]:
round(students_performance[students_performance['lunch'] == 'free/reduced']['math score'].mean(), 0)

59.0

### Задание 9.11
Каков процент студентов, родители которых имеют высшее образование уровня бакалавриата (bachelor's degree)?

In [ ]:
round((students_performance[students_performance['parental level of education'] == "bachelor's degree"].shape[0] / 10) , 0)

12.0

### Задание 9.12
Насколько медианный балл по письму у студентов в расовой группе А отличается от среднего балла по письму у студентов в расовой группе C?

In [ ]:
abs(students_performance[students_performance['race/ethnicity'] == 'group A']['writing score'].median() - students_performance[students_performance['race/ethnicity'] == 'group C']['writing score'].median())

6.0